README.md
# Mooberry Book Manager
**Contributors:** mooberrydreams  
**Tags:** book, author, publisher, writer, books, writing, publishing, authors   
**Requires at least:** 3.8.0  
**Tested up to:** 4.5   
**Stable tag:** 3.1   
**License:** GPLv2 or later  
**License URI:** http://www.gnu.org/licenses/gpl-2.0.html  

## Description
An easy-to-use system for authors. Add your new book to your site in less than 5 minutes.


No coding is necessary to use Mooberry Book Manager. Adding books is as easy as filling out a form. Include as much or as little information as you want, including the book cover, links to purchase the book, reviews of the book, an excerpt, and more.

Mooberry Book Manager will create a page for each book, ensuring a consistent look on all of your pages.

Organize your book into grids with just a few clicks. Grids can include all of your books or a selection of books, and you choose how they are ordered. You can create multiple grids; for example, an "Available Now" page and a "Coming Soon" page. Grids update automatically when you edit or add books. Set it and forget it!

Feature books on your sidebar with four options:
* A random book  
* The newest book  
* A book that's coming soon  
* A specific book  
	
Mooberry Book Manager works with your chosen theme to provide a consistent look throughout your website.

Requires Wordpress 3.8. The admin screens (for creating books, etc.) require Javascript, but the public pages do not.

##Getting Started with Mooberry Book Manager  
After installing and activating Mooberry Book Manager, you'll now have a Books menu.  Use that to add your books to your website.

In order for your books to appear on your website, you need to add a Book Grid and/or a Widget.

To Add a Book Grid:  
* Create a new Page or edit an existing one  
* Scroll down to the Book Grid Settings section  
* Choose the books, grouping, and sort order  
* Save the page and view it. Click on any book cover to get details of the book  

To Add a Widget:  
* Go to Appearance -> Widget  
* Drag the Mooberry Book Manager Book Widget to the Widget Area of your choice  
* Choose the book to display  
* Save the widget and view your website. Click on the book cover to get the details of the book.  

Additonal questions?  
Download the [User Manual](http://www.mooberrydreams.com/support/mooberry-book-manager-support/) (PDF format)

Want regular updates? 
* Like Mooberry Dreams on Facebook: https://www.facebook.com/MooberryDreams
* Follow Mooberry Dreams on Twitter: https://twitter.com/MooberryDreams
* Check out the blog: http://www.mooberrydreams.com/blog
* Subscribe to Mooberry Dreams' mailing list: http://www.mooberrydreams.com/products/mooberry-book-manager/




## Changelog
#### 3.1  
* Fixed: Now compatible with Multisite, including Network Activation  
* Fixed: Migration problems from version 3.0  
* Fixed: Problems with W3 Total Cache   
* Fixed: Not pulling grid default cover height from settings, always defaulting to 200  
* Improved: For new insallations, activation process simplified and sped up (images no longer uploaded to Media Library)  
* Added: Russian Ruble currency  
* Removed: unused images  
* Wordpress 4.5 Compatibility  

#### 3.0 
* Added: Book Grids can now be filtered and grouped by Editors, Illustrators, Cover Artists, and Publishers
* Added: Book Grids can now be grouped by several levels
* Added: Additional pretty links, eg /book/tag/{some-tag} which lists all matching books but display the full book page not a grid
* Added: Ability to update/change "Add to Goodreads" and "Coming Soon" book cover images (on settings page)
* Added: Custom text above and below auto-generated Book Grids (Wordpress 4.4+ only)
* Added: Tip Jar
* Added: Greek Translation
* Added: Indian Rupees (INR/₹) to currency dropdown options
* Improved: Book Cover metabox moved above taxonomies, book info, and buy/download links metaboxes for greater visibility
* Improved: Book Pages and auto-generated Book Grids now created using a shortcode, which should allow them to play nicely with more themes
* Improved: Most book data is now stored in a custom table, allowing for faster and more complex grids
* Improved: URLs for auto-generated Book Grids can now be customized in settings
* Improved: Settings Page moved to main Admin menu for greater visibility and tabs are now sub-pages of the Settings menu
* Improved: New Page Settings sub-page to better categorize settings
* Improved: 2nd set of retailer/download links only displays if excerpt is more than 1500 characters AND excerpt is expanded
* Improved: User interface and instructions/descriptions of fields updated/added
* Improved: "Published on" date now uses the format set in Wordpress General settings
* Improved: Browser tab title is now the title of grid when viewing an auto-generated grid
* Improved: Accessibility improved by all images now using ALT tags set by user OR a default ALT tag in case the user has not set one.
* Improved: Reordered metaboxes on Book Edit page
* Improved: Added filter to move Yoast SEO metabox lower on the Edit Book page
* Improved: Formats & Editions box defaults to closed on Book Edit page. Entering all this info looks intimidating and probably 95% of authors don't want/need this info on the website anyway.
* Improved: Added up/down buttons on repeated fields on Edit Book page to allow re-ordering (reviews, retailer links, download links, and editions)


#### 2.4.2 
* Fixed: Switched from site_url to home_url to create series grid link on book page (caused a problem when Wordpress is installed in a different directory than web URL)  
* Improved: Reduced size of included images. Hopefully will help 500 errors on activation.  


#### 2.4.1
* Fixed: Error about non-object sometimes happening (grid styles)

#### 2.4  
* Added: South African Rand (ZAR/R) to currency dropdown options  
* Added: Ability to update the user who created the book (may need to enable "author" in the Screen Options)  
* Updated: Compliant with Wordpress 4.4 update  

#### 2.3
* New: Added Russian translation
* New: Added Italian translation
* New: Buy Links on the bottom of the book page are now hidden if the excerpt is collapsed
* New: Added Editors, Illustrators, and Cover Artists
* Fixed: Remove roles when uninstalled

#### 2.2  
* New: Added Spanish translation  
* Fixed: Compliant with Wordpress 4.3 (and works for < Wordpress 4.3)  
* Fixed: Retailer images missing since version 2.0 restored  
* Fixed: Improved translation string for Series and Genre grids for more flexibility  

#### 2.1 
* New: Book Grids are now responsive  
* New: Added Serbo–Croatian translation  
* New: Choose which theme template to use for your book pages  
* Improved: Filter Book Grids by tag  
* Improved: Group Book Grids by tag  
* Changed: Book Grids cover height configurable, instead of number of book per row  

#### 2.0.1  
* Flushes rewrite rules  

#### 2.0 
* New: Added General Settings, Publisher Settings, and Edition Formats Settings
* New: Added Editions to book pages 
* New: Added Librarian and Master Librarian Roles
* New: Added French translation
* Improved: Redesigned layout of Book Page
* Improved: Blank data defaults and labels no longer display on Book Page
* Improved: Added Comments support to books
* Fixed: Moved tags to new custom taxonomy instead of using post tags.

For more revision history, see changelog.md



changelog.md

#### 4.2.3 
* Added: Option to turn off SEO META tags

#### 4.2.2 
* Fixed: Sites using SmartCrawl plugin need an override for taxonomy grid pages

#### 4.2.1 
* Fixed: Bug on filter for no books found text on book grids

#### 4.2 
* Added: Option to use book cover as featured image

#### 4.1.22 
* Updated: streamlined arguments to book page filters

#### 4.1.21 
* Added: Vietnam currency
* Added: many filters to the book page

#### 4.1.20 
* Fixed: if the book buy link already contains a ? and has an affiliate link with the affiliate code appended to the link, the affiliate code with start with a & instead of a ?.  Thanks Andre Guirard

#### 4.1.19 
* Fixed: shows summary of book on search page for themes that output post_excerpt
* Fixed: CSS issue on mobile with lists of checklists such as with multi-author

#### 4.1.18 
* Fixed: show taxonomy terms on book grid edit page even if they have no books

#### 4.1.17 
* Fixed: third time is the charm? Finishing issues on grids with books/no books

#### 4.1.16 
* Fixed: Closed off another path of returning all books when it should be no books

#### 4.1.15 
* Fixed: Be sure not to return ALL books on a grid that is really returning NO books
* Fixed: Display currency with price in editions

#### 4.1.14 
* Fixed: obscure PHP notice on some settings configurations

#### 4.1.13 
* Updated: iBooks updated to Apple Books in list of Retailers that come pre-installed with MBM


#### 4.1.12 
* Fixed: Bug in logic to remove genre, series, and tags from quick/bulk edit in WP5.1+ was bringing them back in WP5.2
* Fixed: Allow MBM3.0 data migration to happen in MBM4.0+ for users who have not kept plugin updated

#### 4.1.11 
* Fixed: Removed genre, series, and tags from quick/bulk edit for WP version 5.1 and higher, due to possible bug in WP 5.1 causing the fields to empty out when saved

#### 4.1.10 
* Fixed: A change in WP 5.1 prevented tags, genres, series, illustrators, editors, and cover artists to not save correctly

#### 4.1.9 
* Fixed: Error causing affiliate code to output twice in bottom list of links on book page

#### 4.1.8 
* Fixed: Removed HTML tags from meta tags (SEO)

#### 4.1.7 
* Added: Support for new Multi-Author feature

#### 4.1.6 
* Fixed: SQL bug relating to Multi-Author

#### 4.1.5 
* Fixed: Version 4.0.19 erroneously removed Publisher as a group by option in book grids

#### 4.1.4 
* Fixed: Issue with book with future publication date not showing date on book page

#### 4.1.3 
* Fixed: Previous update erroneously removed Publisher as an option in book grids
* Fixed: Check for object_id in menu object just in case of odd theme or plugin conflicting?
* Fixed: Fix bug that wasn't setting MBM Tax Grid Page on activation

#### 4.1.2 
* Updated: Internal changes to support future features

#### 4.1.1 
* Fixed: Fatal error: Declaration of CMB2_Type_File::render() must be compatible with CMB2_Type_Text::render($args = Array)  on Settings-> Book Page. (caused by error in included library CMB2 when used with PHP 7.2 )
* Added: Brazilian Portuguese translations ( thanks Vinicius Cubas Brand! )

#### 4.1 
* Updated: Improved mobile layout for book pages
* New: Compatible with Duplicate Post so that books can be duplicated into new posts

#### 4.0.21 
* Fixed: Compatible with Yoast SEO to use the book's summary as content for analyzing

#### 4.0.20 
* Fixed: bugs when importing/exporting books with publishers

#### 4.0.19 
* Updated: Compliant with WP 4.9: allows shortcodes in widgets without using global $post
* Various bugs/refactoring

#### 4.0.18 
* Issues with previous update being recognized by wordpress

#### 4.0.17 
* Fixed: bug in grid book image

#### 4.0.16 
* Fixed: Add comment support to books
* Fixed: Add support for quick/bulk edit for extensions

#### 4.0.15 
* Fixed: Output paragraph and styling on tax grid descriptions
* Fixed: Display H1-H6 headings in the summary as inline
* Fixed: Tax Grid page creation no longer broken
* Fixed: Backwards compatibility issues when upgrading from very old versions

#### 4.0.14 
* Added: Compatibility with future updates

#### 4.0.13 
* Fixed: Allow post types allowed to have book grid shortcode button to be customized

#### 4.0.12 
* Fixed: Retailers, Publishers, and Formats now display in alphabetical order in drop down lists
* Fixed: Series, Tags, Genres, Cover Artists, Editors, and Illustrators now included in Book Grid selection even if no books have been added to them yet
* Fixed: search issue

#### 4.0.11 
* Fixed: Bug with backwards compatibility with MA

#### 4.0.10 
* Changes to support upcoming features

#### 4.0.9 
* Fixed: Problem with author sorting on grids with Multi-Author ( requires Multi-Author v1.12+ )

#### 4.0.8 
* Fixed: Permissions for Illustrators

#### 4.0.7 
* Fixed: Issue with affiliate code not using before position properly

#### 4.0.6 
* Fixed: Issue caused by version 4.0.5 preventing books from being saved

#### 4.0.5 
* Fixed: Issue causing widgets to be removed. Widgets may still be removed after THIS update, however they should not be going forward. I apologize for the inconvenience

#### 4.0.4 
* Fixed: error on License Activation screen when there's only one extension installed
* Fixed: only return 1 book for book widget when used with multi-author extension

#### 4.0.3 
* Fixed: Truncate meta description around 200 chars
* Fixed: Add description, open graph, and twitter card meta tags to taxonomy grids

#### 4.0.2 
* Fixed: don't use cropped images in grid
* Fixed: allow shortcodes in before/after grid descriptions
* Fixed: fixed code to migrate tax grid before/after descriptions

#### 4.0.1 
* Fixed: bugs

#### 4.0 
* Fixed: Edition size (width/height) no longer rounds to nearest whole number
* Added: Ability to import and export books (for backup and site migration purposes)
* Added: Open-Graph meta tags compatible with FB for sharing. Adds book cover, title,  summary, and ISBN. 

book-widget.php
<?php
/**
 *  This file does the book widget  
 *  
 */
 
 
class mbdb_book_widget extends WP_Widget {

	// constructor
	// 3.1 -- added customize_selected_refresh for WP 4.5
	function __construct() {
		$widget_ops = array('classname' => 'mbdb_book_widget', 
							'description' => __('Shows the cover of the book of your choosing with a link to the book page', 'mooberry-book-manager'),
							 'customize_selective_refresh' => true,);
		
		parent::__construct('mbdb_book_widget', 'Mooberry Book Manager '. _x('Book', 'noun', 'mooberry-book-manager'), $widget_ops  );

	}

	// widget form creation
	function form($instance) {	
		$instance = wp_parse_args((array) $instance);
		// Check values
		if( $instance) {
			 do_action('mbdb_widget_pre_get_data', $instance);
			 $mbdb_widget_title = esc_attr($instance['mbdb_widget_title']);
			 $mbdb_widget_show_title = esc_attr($instance['mbdb_widget_show_title']);
			 $mbdb_widget_type = esc_attr($instance['mbdb_widget_type']);
			 $mbdb_bookID = esc_attr($instance['mbdb_bookID']);
			 $mbdb_widget_cover_size = esc_attr($instance['mbdb_widget_cover_size']);
			 do_action('mbdb_widget_post_get_data', $instance);
		} else {
			 do_action('mbdb_widget_pre_set_defaults'); 
			 $mbdb_widget_title = '';
			 $mbdb_bookID=0;		
			 $mbdb_widget_show_title = 'yes';
			 $mbdb_widget_type = 'random';
			 $mbdb_widget_cover_size = 100;
			 $mbdb_book_title = '';
			 do_action('mbdb_widget_post_set_defaults'); 
		}
		
		$options = apply_filters('mbdb_book_widget_options', array('random' => __('Random Book', 'mooberry-book-manager'),
							'newest'	=>	__('Newest Book', 'mooberry-book-manager'),
							'coming-soon'	=>	__('Future Book', 'mooberry-book-manager'),
							'specific'	=>	__('Specific Book', 'mooberry-book-manager')
						)
					);
							
		$widget_type_dropdown = mbdb_dropdown($this->get_field_id('mbdb_widget_type'), $options, $mbdb_widget_type, 'no', 0, $this->get_field_name('mbdb_widget_type'));
		include( plugin_dir_path(__FILE__)  . '/views/admin-book-widget.php');
	}

	// widget update
	function update($new_instance, $old_instance) {
		$instance = $old_instance;
		do_action('mbdb_widget_pre_update', $new_instance, $old_instance);
		
		$instance['mbdb_widget_title'] = strip_tags($new_instance['mbdb_widget_title']);
		
		$instance['mbdb_bookID'] = strip_tags($new_instance['mbdb_bookID']);
		
		$instance['mbdb_widget_show_title'] = strip_tags($new_instance['mbdb_widget_show_title']);
		
		$instance['mbdb_widget_type'] = strip_tags($new_instance['mbdb_widget_type']);
		
		if ($new_instance['mbdb_widget_cover_size'] < 50) {
			$new_instance['mbdb_widget_cover_size'] = 50;
		}
		
		$instance['mbdb_widget_cover_size'] = strip_tags($new_instance['mbdb_widget_cover_size']);
		
		do_action('mbdb_widget_post_update', $new_instance, $instance);
		
		return apply_filters('mbdb_widget_update', $instance, $new_instance);
	}

	// widget display
	function widget($args, $instance) {
		extract($args);
		
		$mbdb_bookID  = $instance['mbdb_bookID'];
		
		$mbdb_widget_type = apply_filters('mbdb_widget_type', $instance['mbdb_widget_type']);
		
	
		$book = null;
		do_action('mbdb_widget_pre_get_books', $instance);
			

		switch ($mbdb_widget_type) {
			case 'random':
				// get book ID of a random book
				$book = apply_filters('mbdb_widget_random_book_list', MBDB()->books->get_random_book(), $instance);
				break;
		
			case "newest":
				// get book ID of most recent book
				
				$book = apply_filters('mbdb_widget_newest_book_list', MBDB()->books->get_most_recent_book(), $instance);
				break;
			
			case "coming-soon":
				// get books with future or blank release dates
				$book = apply_filters('mbdb_widget_coming_soon_book_list', MBDB()->books->get_upcoming_book(), $instance); 
				break;
			
			case "specific":
				// make sure seected book is still a valid book
				$book = apply_filters('mbdb_widget_specific_book_list', MBDB()->books->get($mbdb_bookID), $instance); 				
				break;
		}
		$book = apply_filters('mbdb_widget_book', $book, $instance);
		
		do_action('mbdb_widget_post_get_books', $instance, $book);
		
		//output
		if ($book == null) {
			$mbdb_bookID = 0;
			$mbdb_book_title = '';
		} else {
			$mbdb_bookID = $book->book_id;
			$mbdb_book_title = get_the_title($mbdb_bookID);
		}
		
		$mbdb_widget_title = apply_filters('mbdb_widget_title', $instance['mbdb_widget_title']);
		$mbdb_widget_show_title = apply_filters('mbdb_widget_show_title', $instance['mbdb_widget_show_title']);
		$mbdb_cover_size = apply_filters('mbdb_widget_cover_size', $instance['mbdb_widget_cover_size']);
		$mbdb_bookID = apply_filters('mbdb_widget_bookID', $mbdb_bookID);
		$mbdb_book_title = apply_filters('mbdb_widget_book_title', $mbdb_book_title);
		
		do_action('mbdb_widget_pre_display');
		echo $before_widget;
		
		echo $before_title . esc_html($mbdb_widget_title) . $after_title;
		
		if ($mbdb_bookID == 0) {
			echo apply_filters('mbdb_widget_no_books_found', '<em>' . __('No books found', 'mooberry-book-manager') . '</em>');
		} else {
			$image_src = $book->cover;
			$image_id = $book->cover_id;
			//$image_src = get_post_meta( $mbdb_bookID, '_mbdb_cover', true );
			
			$book_link = get_permalink( $mbdb_bookID );
			
			if ( $book_link != '' ) {
				do_action('mbdb_widget_pre_link', $book_link);
				echo '<A class="mbm-widget-link" HREF="' . esc_url($book_link) . '"> ';
			}
			
			$image_src = mbdb_get_cover( $image_src, 'widget' );
			
			if ($image_src && $image_src !== '') {
				do_action('mbdb_widget_pre_image', $image_src);
				$alt = mbdb_get_alt_text( $image_id, __('Book Cover:', 'mooberry-book-manager') . ' ' . $mbdb_book_title);
				echo '<img class="mbm-widget-cover" style="width:' . esc_attr($mbdb_cover_size) . 'px;padding-top:10px;" src="' . esc_url($image_src) . '" ' . $alt . '  /> ';
				do_action('mbdb_widget_post_image', $image_src);
			}
			if ($mbdb_widget_show_title == 'yes') { 
				if ( $mbdb_book_title != '' ) {
					do_action('mbdb_widget_pre_book_title', $mbdb_book_title);
					echo '<P><span class="mbm-widget-title">' . esc_html($mbdb_book_title) . '</span></P>';
					do_action('mbdb_widget_post_book_title', $mbdb_book_title);
				}
			}
			if ($book_link != '') {
				echo '</A>';
				do_action('mbdb_widget_post_link');
			}
			
		}
		//echo '</div>' . 
		echo $after_widget;
		do_action('mbdb_widget_post_display');
	}
	
}
